## PR Review Summary

**Status:** Request Changes

### Strategic-Fit Decision

Per §9 Strategic-Fit Step-Back:
- **Decision:** Request Changes
- **Rationale:** The main service migration shape is correct, but the PR leaves a tracked neural-link helper importing the removed `services/` directory. That is a concrete post-migration breakage, not a style concern. The PR body also leaves a stale `<commit-sha>` placeholder and does not satisfy #10994 AC12's requested diff summary table, so the public merge artifact is not yet in sync with the ticket contract.

**Peer-Review Opening:** The core NL migration is close: the 9-service move, preserved `NeuralLink_*` aliases, `Server.mjs` boot-order import updates, and moved RecorderService spec all line up with the M6 pattern. The blockers below are narrow and should be quick to repair.

---

### Context & Graph Linking

- **Target Epic / Issue ID:** Resolves #10994
- **Related Graph Nodes:** #10986 M6 Tier-1 flat SDK epic, #10991 KB migration, #10993 GH-WF migration, #10975 NL BaseServer/custom boot-order precedent, `ai/services.mjs`, `ai/services/neural-link/*`, `ai/mcp/server/neural-link/Server.mjs`.

---

### Depth Floor

**Challenge (per guide §7.1):** The diff grep surfaced a leftover tracked helper outside the moved service tree: `ai/mcp/server/neural-link/test_connection_service.mjs` still imports `./services/ConnectionService.mjs`. Since the PR removes that directory, the helper now crashes on import.

Empirical check run locally at head `5197f99ce596a738c498b1bd89dea8a2e0537ee0`:

```bash
node -e "import('./ai/mcp/server/neural-link/test_connection_service.mjs').catch(e => { console.error(e.message); process.exit(1); })"
```

Result:

```text
Cannot find module '/Users/Shared/codex/neomjs/neo/ai/mcp/server/neural-link/services/ConnectionService.mjs' imported from /Users/Shared/codex/neomjs/neo/ai/mcp/server/neural-link/test_connection_service.mjs
```

**Rhetorical-Drift Audit (per guide §7.4):**

- [ ] PR description claims all consumers work unchanged via aliases, but a tracked NL helper still consumes the old physical service path.
- [x] Anchor & Echo summaries in touched source use precise codebase terminology; no metaphor drift observed.
- [x] No `[RETROSPECTIVE]` tag inflation in PR body.
- [x] Linked anchors #10975 / #10986 / sibling PRs are mechanically relevant.

**Findings:** Request Changes for the stale helper import + PR-body contract drift.

---

### Graph Ingestion Notes

- **`[KB_GAP]`**: N/A.
- **`[TOOLING_GAP]`**: `gh pr view` / `gh pr checks` briefly failed with GitHub connection errors during review; retry with escalation succeeded. Also, direct `node -e "await import('./ai/services/neural-link/toolService.mjs')"` fails without Neo bootstrap (`Neo is not defined`), while `ai/services.mjs` import succeeds. I am not treating that as a PR blocker because this service is normally loaded through the MCP/SDK bootstrap, but it is worth remembering when choosing smoke-test shapes.
- **`[RETROSPECTIVE]`**: For flat SDK migrations, grep for old physical `ai/mcp/server/<server>/services` paths must include scripts/helpers under the server root, not only `Server.mjs`, tests, and `ai/services.mjs`.

---

### Provenance Audit

Internal origin: M6 flat SDK migration under #10986, with NL-specific custom boot-order constraints from #10975. No external-framework port observed.

---

### Close-Target Audit

- [x] Close-targets identified: #10994.
- [x] #10994 labels checked live: `enhancement`, `ai`, `refactoring`, `architecture`, `release:v13`; not `epic`.

**Findings:** Pass.

---

### Contract Completeness Audit

- [x] #10994 contains a Contract Ledger matrix.
- [ ] Implemented PR diff matches the ledger exactly.

**Findings:** Contract drift:
- AC4 / AC5 drift: a tracked helper still imports the removed old service path, so not all NL consumer/helper code works post-migration.
- AC12 drift: PR description asks for a brief diff summary table, but the body currently uses prose/bullets and leaves the commit section as `<commit-sha>`.

---

### Evidence Audit

- [x] PR body contains an `Evidence:` declaration line.
- [x] Focused local unit evidence was independently re-run by reviewer and passed.
- [ ] Evidence misses the stale tracked helper import because that helper is outside the moved unit-spec path.
- [x] No L1/L2 evidence was promoted to L3/L4 framing.

**Findings:** Evidence class is appropriate, but the uncovered helper import gap blocks approval.

---

### Source-of-Authority Audit

N/A — no operator or peer authority citation is used as the basis for a demand.

---

### MCP-Tool-Description Budget Audit

N/A — PR does not touch `ai/mcp/server/*/openapi.yaml` descriptions.

---

### Wire-Format Compatibility Audit

No JSON-RPC payload or tool schema changes observed. The `toolService.mjs` openapi path and `getCurrentTurnId` import path were audited; path math is correct.

**Findings:** Pass.

---

### Cross-Skill Integration Audit

- [x] No skill files or workflow conventions changed.
- [x] No new MCP tool added.
- [x] New architectural convention is inherited from #10986 M6 and documented in the ticket/PR body.
- [x] Downstream SDK alias consumers are represented through `ai/services.mjs` exports.

**Findings:** Pass, except for the stale tracked helper path captured under Contract Completeness / Required Actions.

---

### Test-Execution & Location Audit

- [x] Branch checked out locally at `5197f99ce596a738c498b1bd89dea8a2e0537ee0`.
- [x] Moved unit spec location is canonical for right-hemisphere service tests: `test/playwright/unit/ai/services/neural-link/RecorderService.spec.mjs`.
- [x] Related tests run:

```bash
npm run test-unit -- test/playwright/unit/ai/services/neural-link/ test/playwright/unit/ai/mcp/server/neural-link/logger.spec.mjs
```

Result: `7 passed (723ms)`.

- [x] SDK alias smoke via `ai/services.mjs` succeeded for sampled NL aliases.
- [ ] Tracked helper smoke failed due stale import as documented above.

**Findings:** Related Playwright tests pass; helper import failure blocks approval.

---

### CI / Security Checks Audit

`gh pr checks 10998` after retry:

- CodeQL: pass
- Analyze (javascript): pass
- unit: pass
- integration-unified: pending at review time

**Findings:** Approval is blocked regardless of pending integration because of the stale helper import. If integration later fails, that will need a separate fix.

---

### Required Actions

To proceed with merging, please address the following:

- [ ] Update or remove `ai/mcp/server/neural-link/test_connection_service.mjs:3`. It currently imports `./services/ConnectionService.mjs`, but that directory is removed by this PR. If the helper remains, it should import from the new SDK path (`../../../services/neural-link/ConnectionService.mjs`) or an equivalent correct path. Re-run the `node -e import(...)` smoke after fixing.
- [ ] Refresh the PR body metadata: replace `<commit-sha>` with the actual commit hash and add the brief diff summary table requested by #10994 AC12, or explicitly explain why the current prose is the intended AC12 equivalent.

---

### Evaluation Metrics

- **`[ARCH_ALIGNMENT]`**: 90 — 10 points deducted because the physical move pattern is correct, but the stale helper path shows the migration sweep did not cover every tracked server-root consumer.
- **`[CONTENT_COMPLETENESS]`**: 85 — 15 points deducted for stale PR body commit metadata and missing/implicit AC12 diff-summary table.
- **`[EXECUTION_QUALITY]`**: 70 — 30 points deducted because the focused Playwright specs pass, but a tracked helper now fails at module resolution after the old directory removal.
- **`[PRODUCTIVITY]`**: 85 — 15 points deducted because the primary 9-service migration is implemented, but AC4/AC5/AC12 are not fully satisfied yet.
- **`[IMPACT]`**: 80 — Major M6 migration step that preserves the SDK alias contract while moving NL services to the flat boundary.
- **`[COMPLEXITY]`**: 45 — Moderate: mostly mechanical file moves and import path updates, with added risk from NL custom `boot()` and `toolService.mjs` path math.
- **`[EFFORT_PROFILE]`**: Quick Win — the remaining fixes are narrow; once repaired, this should unblock the next M6 lane without broader redesign.

Closing: Request Changes is intentionally narrow here. The main architecture looks right; fix the stale helper import and PR-body contract drift, then this should be a fast re-review.

--- a/ai/mcp/server/neural-link/test_connection_service.mjs
+++ b/ai/mcp/server/neural-link/test_connection_service.mjs
@@ -1,6 +1,6 @@
 import Neo from '../../../../src/Neo.mjs';
 import * as core from '../../../../src/core/_export.mjs';
-import ConnectionService from './services/ConnectionService.mjs';
+import ConnectionService from '../../../services/neural-link/ConnectionService.mjs';
 
 console.log('ConnectionService imported');
 console.log('Is instance?', ConnectionService instanceof Neo.core.Base);

